### PR TITLE
GCR vulnerability scanning

### PR DESCRIPTION
**Pull request by grosser**
/fyi @jonmoter 

 - warn users about vulnerabilities or stop deploys if they opt-in
 - updating our cache store to mem_cache_store since dalli does not support race_condition_ttl and mem_cache_store seems to be the future as per discussion in https://github.com/petergoldstein/dalli/issues/286

screen shot 2018-01-31 at 3 24 42 pm

screen shot 2018-01-31 at 3 26 03 pm

screen shot 2018-01-31 at 3 28 09 pm

screen shot 2018-02-01 at 12 32 11 pm

 - the vuln scan shows "success" for scans that are in-progress, so we need to either force a wait or talk to google folks ... even the CLI shows not difference (need to try with a build that just finished)

### CODE REVIEW

**Comment by grosser** [comment]
@jonmoter @ragurney

File: plugins/gcloud/lib/samson_gcloud/image_scanner.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+module SamsonGcloud
+  class ImageScanner
+    WAITING = 0
+    SUCCESS = 1
+    FOUND = 2
+    ERROR = 3
+
+    GRACE_PERIOD = 10.minutes # time to wait for scan to complete since we cannot tell pending from success
```
Comment (ragurney) [inline plugins/gcloud/lib/samson_gcloud/image_scanner.rb]: todo

File: plugins/gcloud/lib/samson_gcloud/samson_plugin.rb [hunk 2]
```diff
@@ -33,3 +36,28 @@ def account
 Samson::Hooks.callback :project_permitted_params do
   :build_with_gcb
 end
+
+Samson::Hooks.callback :stage_permitted_params do
+  :block_on_gcr_vulnerabilities
+end
+
+
```
Comment (ragurney) [inline plugins/gcloud/lib/samson_gcloud/samson_plugin.rb]: newline

File: plugins/gcloud/lib/samson_gcloud/samson_plugin.rb [hunk 3]
```diff
@@ -33,3 +36,28 @@ def account
...
+Samson::Hooks.callback :ensure_build_is_successful do |build, job, output|
+  if ENV['GCLOUD_IMAGE_SCANNER']
+    status = build.gcr_vulnerabilities_status_id
+    unless [SamsonGcloud::ImageScanner::SUCCESS, SamsonGcloud::ImageScanner::FOUND].include?(status)
```
Comment (ragurney) [inline plugins/gcloud/lib/samson_gcloud/samson_plugin.rb]: finished

File: plugins/gcloud/lib/samson_gcloud/samson_plugin.rb [hunk 4]
```diff
@@ -33,3 +36,28 @@ def account
...
+      status = SamsonGcloud::ImageScanner.scan(build)
+    end
+
+    success = (status == SamsonGcloud::ImageScanner::SUCCESS)
+    build.update_attributes!(gcr_vulnerabilities_status_id: status)
+    message = SamsonGcloud::ImageScanner.status(status)
+    message += ", see #{SamsonGcloud::ImageScanner.result_url(build)}" unless success
+
+    if success || !job.deploy.stage.block_on_gcr_vulnerabilities
+      output.puts message
```
Comment (ragurney) [inline plugins/gcloud/lib/samson_gcloud/samson_plugin.rb]: returm true/false

File: plugins/gcloud/lib/samson_gcloud/image_scanner.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    FINISHED = [SUCCESS, FOUND].freeze
+
+    # time to wait for scan to complete since we cannot tell pending from success
+    # TODO: find a way to actually tell success from pending
+    GRACE_PERIOD = 10.minutes
+
+    class << self
+      # found by inspecting
+      # gcloud alpha container images list-tags gcr.io/$PROJECT_ID/base/alpine --show-occurrences --log-http
+      def scan(build)
```
Comment (jonmoter) [inline plugins/gcloud/lib/samson_gcloud/image_scanner.rb]: Shall we cache the results of this for a while, so we don't re-fetch the same information on every page load?

Comment (grosser) [reply plugins/gcloud/lib/samson_gcloud/image_scanner.rb]: the result is cached on the build in a column

**Pull request merged**